src/context/directory/handlers/clientGrants.ts

```diff
@@ -39,15 +40,43 @@ async function dump(context: DirectoryContext): Promise<void> {
   const grantsFolder = path.join(context.filePath, constants.CLIENTS_GRANTS_DIRECTORY);
   fs.ensureDirSync(grantsFolder);
 
+  const allResourceServers = await context.mgmtClient.resourceServers.getAll({
+    paginate: true,
+  });
+
+  const allClients = await context.mgmtClient.clients.getAll({
+    paginate: true,
+  });
```

Comment:
What do you think if we optimize the algorithm a little bit?

1. Fetch all client grants - `GET https://{{auth0_domain}}/api/v2/client-grants`
2. While looping through each client grant
3. Check if we already dumped a client and find it based on the ID, otherwise find the client name using the id and `GET https://{{auth0_domain}}/api/v2/clients/:id`
4. Check if we already dumped a resource server and find it based on the audience, otherwise make a GET request using the path escaped audience as the `:id` e.g. `GET https://{{auth0_domain}}/api/v2/resource-servers/https:%2F%2Fmy-tenant.eu.auth0.com%2Fapi%2Fv2%2F` (this will actually work).

Reply:
Alternatively, to optimize for tenants with large amounts of clients and resource servers, we could fetch all of them on lines 43 and 47 but set as well the `per_page` property to 100 (max).

Reply:
You and I talked this over earlier. To recap, several performance considerations needed to be made, particularly in the case of tenants with hundreds of clients and/or resource servers. In those cases, it is likely that tenants also have many client grants too, so while making one batched call of all the clients and resource servers upfront, at least that half-dozen or so HTTP requests will be the presumed maximum. This is sub-optimal for smaller tenants which have fewer clients, client grants and resource servers, but not practically significant for a tool that makes hundreds of other HTTP requests.

>Alternatively, to optimize for tenants with large amounts of clients and resource servers, we could fetch all of them on lines 43 and 47 but set as well the per_page property to 100 (max).

I looked into this and the default for pagination of this tool is 100, so no need to specify so explicitly.

Reply:
After our conversation, I do realize that there is one important optimization that can be made – returning early if there are zero client grants. Otherwise, the client grants and resource servers will still be fetched despite no client grants putting that data to use.